src/config/field-overrides/relationshipType.js

```diff
@@ -0,0 +1,141 @@
+
+const TRACKED_ENTITY_INSTANCE = 'TRACKED_ENTITY_INSTANCE';
+const PROGRAM_INSTANCE = 'PROGRAM_INSTANCE';
+const PROGRAM_STAGE_INSTANCE = 'PROGRAM_STAGE_INSTANCE';
+
+const constraintOptions = [TRACKED_ENTITY_INSTANCE, PROGRAM_INSTANCE, PROGRAM_STAGE_INSTANCE];
+
+// Map of the different names of the embedded objects, according to selected constraint-type
+const constraintToObjectType = {
+    TRACKED_ENTITY_INSTANCE: 'trackedEntityType',
+    PROGRAM_INSTANCE: 'program',
+    PROGRAM_STAGE_INSTANCE: 'programStage',
+};
+
+const resolveToTrue = () => true;
+
+class Constraint extends Component {
+    constructor(props, context) {
+        super(props);
+        this.translate = context.d2.i18n.getTranslation.bind(context.d2.i18n);
+
+        let embeddedValue = null;
+        let relationshipEntity = null;
+
+        if (props.value) {
+            relationshipEntity = props.value.relationshipEntity;
+            if (relationshipEntity) {
+                const objectName = constraintToObjectType[relationshipEntity];
+                embeddedValue = props.value[objectName];
+            }
+        }
+
+        const label = this.getRelationshipEntityLabel(relationshipEntity);
+
+        this.state = {
+            searchResults: [],
+            relationshipEntity,
+            selectLabel: label,
+            embeddedValue,
+        };
+    }
+
+    getRelationshipEntityLabel = relationshipEntity => {
+        switch (relationshipEntity) {
+            case TRACKED_ENTITY_INSTANCE:
+                return this.translate('tracked_entity_type');
+            case PROGRAM_INSTANCE:
+                return this.translate('program');
+            case PROGRAM_STAGE_INSTANCE:
+                return this.translate('program_stage');
+        }
+
+        return null;
+    };
+
+    selectRelationshipEntity = (_, __, value) => {
+        const selectLabel = this.getRelationshipEntityLabel(value);
+
+        this.setState({
+            relationshipEntity: value,
+            selectLabel,
+        });
+    };
+
+    handleEmbeddedValueSelect = ({ target: { value } }) => {
+        const objType = constraintToObjectType[this.state.relationshipEntity];
+        const relationshipConstraint = {
+            relationshipEntity: this.state.relationshipEntity,
+            [objType]: {
+                id: (value && value.id) || null,
+            },
+        };
+        this.props.onChange({
+            target: {
+                value: relationshipConstraint,
+            },
+        });
+
+        this.setState({
+            embeddedValue: relationshipConstraint[objType],
+        });
+    };
+
+    render = () => {
+        return (
+            <div>
+                <SelectField
+                    fullWidth
+                    value={this.state.relationshipEntity}
+                    onChange={this.selectRelationshipEntity}
+                    floatingLabelText={this.translate('from_constraint')}
```

Comment:
This needs to by dynamic so it can display either "From constrain" or "To constraint"

Reply:
Thanks for noticing!